ngraph/core/reference/include/ngraph/runtime/reference/convolution.hpp

```diff
@@ -24,338 +25,340 @@
-
-                    size_t batch_index = out_coord[out_batch_axis];
-                    size_t out_channel = out_coord[out_channel_axis];
+                    T buffer;
+                    const Shape shape;
+                    Tensor(T buf, Shape s)
+                        : buffer{buf}
+                        , shape{s} {};
+                };
+
+                enum class ConvolutionType
+                {
+                    Conv1D,
+                    Conv2D,
+                    Conv3D
+                };
 
-                    // For the in we need to iterate the coordinate:
-                    //
-                    //   I:
-                    //
-                    // over the range (noninclusive on the right):
-                    //
-                    //   (N,0,s_1*i_1,s_2*i_2,...,s_n*i_n) ->
-                    //
-                    //     (N+1,
-                    //      chans_in_count,
-                    //      s_1*i_1+ l_1*filter_dims_1,
-                    ///       ...,
-                    ///     s_n*i_n +l_n*filter_dims_n)
-                    //
-                    // with strides:
-                    //
-                    //   (1,l_1,...,l_n).
-                    //
-                    // Note that we are iterating within the *padded* and *dilated* in batch, so
-                    // further down we must check the current coordinate is in the pad or dilation
-                    // gap.
+                struct ConvolutionParams
+                {
+                    std::vector<int> strides;
+                    std::vector<int> dilation;
+                    std::vector<int> pads_begin;
+                    std::vector<int> pads_end;
+
+                    ConvolutionParams(const Strides& strides_,
+                                      const Strides& dilation_,
+                                      const CoordinateDiff& pads_begin_,
+                                      const CoordinateDiff& pads_end_)
+                        : strides{strides_.begin(), strides_.end()}
+                        , dilation{dilation_.begin(), dilation_.end()}
+                        , pads_begin{pads_begin_.begin(), pads_begin_.end()}
+                        , pads_end{pads_end_.begin(), pads_end_.end()} {};
+                };
+
+                size_t compute_size(ngraph::Shape shape, size_t start_axis)
+                {
+                    size_t size = 1;
+                    for (size_t i = start_axis; i < shape.size(); i++)
+                    {
+                        size *= shape[i];
+                    }
+                    return size;
+                }
 
-                    size_t n_spatial_dimensions = in_shape.size() - 2;
-                    size_t n_in_channels = in_shape[in_channel_axis];
+                template <typename INPUT, typename FILTER, typename OUTPUT, typename ACCU>
+                void convolve_1D_channels(const ConvolutionParams& p,
```

Comment:
I'm nitpicking a bit, and this isn't super important, but I'd prefer the variables you're using to iterate to be more clear on the first glance, so filter_x, input_x, channel_idx and so on. This can be deduced by the reader but the code is a bit scary on the first look

Reply:
I'd also prefer more descriptive names but the 100 line limit set for the formater gets code splitted into multiple lines and then code is even less readbile. So, in my mind I've chosen lesser evil - short index names.